### ISSUE TITLE
[Good First Issue][NNCF]: Remove backend-specific methods from the common layer attributes

### ISSUE THREAD

**Issue by daniil-lyakhov** [openvinotoolkit/nncf#3249]
### Context

Greetings, fellow scientists😺!

Recently we've discovered an old code  that becomes confusing when used with the latest algorithms:

https://github.com/openvinotoolkit/nncf/blob/develop/nncf/common/graph/layer_attributes.py#L56-L79
https://github.com/openvinotoolkit/nncf/blob/develop/nncf/common/graph/layer_attributes.py#L135-L136

the problematic methods are `get_weights_shape`, `get_target_dim_for_compression` and `get_bias_shape` . Current algorithms use NNCFGraph to retrieve correct info, and not every backend follow the hardcoded weights shape layout / compression dim.

### What needs to be done?

The task is to create the following functions in the https://github.com/openvinotoolkit/nncf/blob/develop/nncf/common/graph/utils.py file:

* `get_weights_shape_legacy(layer_attributes: WeightedLayerAttributes)` 
* `get_target_dim_for_compression_legacy(layer_attributes: WeightedLayerAttributes)` 
* `get_bias_shape_legacy(layer_attributes: WeightedLayerAttributes)`

Additionally, all calls to the following methods should be replaced with their corresponding legacy function calls:

* `get_weights_shape` -> `get_weights_shape_legacy`
* `get_target_dim_for_compression` -> `get_target_dim_for_compression_legacy`
* `get_bias_shape` -> `get_bias_shape_legacy`

Furthermore, the docstrings of these methods should clarify that they are valid only for PyTorch and TensorFlow backends.

### Example Pull Requests

https://github.com/openvinotoolkit/nncf/pull/3179

### Resources

- [Contribution guide - start here!](https://github.com/openvinotoolkit/nncf/blob/develop/CONTRIBUTING.md)
- [Intel DevHub Discord channel](https://discord.gg/7pVRxUwdWG) - engage in discussions, ask questions and talk to OpenVINO developers
- [How to link your Pull Request to an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#manually-linking-a-pull-request-to-an-issue-using-the-pull-request-sidebar)

### Contact points

@daniil-lyakhov

**Comment by shumaari** [comment]
.take

**Comment by shumaari** [comment]
I am still looking into the issue and would like to clarify my understanding:

### Function calls for three methods need to be replaced by their legacy versions 

`layer_attributes.get_weight_shape` -> `layer_attributes.get_weight_shape_legacy`
`layer_attributes.get_target_dim_for_compression` -> `layer_attributes.get_target_dim_for_compression_legacy`
`layer_attributes.get_bias_shape` -> `layer_attributes.get_bias_shape_legacy`


##### Possible locations for `layer_attributes.get_weight_shape`



https://github.com/openvinotoolkit/nncf/blob/develop/nncf\experimental\tensorflow\quantization\algorithm.py#L136
https://github.com/openvinotoolkit/nncf/blob/develop/nncf\experimental\torch\sparsity\movement\layers.py#L170
https://github.com/openvinotoolkit/nncf/blob/develop/nncf\experimental\torch\sparsity\movement\structured_mask_handler.py#L212
https://github.com/openvinotoolkit/nncf/blob/develop/nncf\torch\quantization\algo.py#L776
https://github.com/openvinotoolkit/nncf/blob/develop/nncf\torch\quantization\algo.py#L1185
https://github.com/openvinotoolkit/nncf/blob/develop/nncf\torch\quantization\init_range.py#L229
https://github.com/openvinotoolkit/nncf/blob/develop/nncf\torch\sparsity\const\algo.py#L29
https://github.com/openvinotoolkit/nncf/blob/develop/nncf\torch\sparsity\magnitude\algo.py#L47
https://github.com/openvinotoolkit/nncf/blob/develop/nncf\torch\sparsity\rb\algo.py#L47

##### Possible locations for `layer_attributes.get_target_dim_for_compression`

https://github.com/openvinotoolkit/nncf/blob/develop/nncf\experimental\common\pruning\nodes_grouping.py#L79
https://github.com/openvinotoolkit/nncf/blob/develop/nncf\torch\quantization\algo.py#L779
https://github.com/openvinotoolkit/nncf/blob/develop/nncf\torch\quantization\init_range.py#L230

##### Possible locations for `layer_attributes.get_bias_shape`

https://github.com/openvinotoolkit/nncf/blob/develop/nncf\experimental\torch\sparsity\movement\layers.py#L188
https://github.com/openvinotoolkit/nncf/blob/develop/nncf\experimental\torch\sparsity\movement\structured_mask_handler.py#L214

### Create the three legacy methods with docstrings

I am still unclear about where these functions could be created? Within class definitions for `WeightedLayerAttributes` and `LinearLayerAttributes` classes in https://github.com/openvinotoolkit/nncf/blob/develop/nncf/common/graph/layer_attributes.py like below or somewhere else? 

![Image](https://github.com/user-attachments/assets/a4388731-a6be-4070-9801-9fd415e6ec17)

I would also like to mention that `get_weight_shape_legacy` and `get_target_dim_for_compression_legacy` methods in `WeightedLayerAttributes` class are mere placeholders. The method which caused the reported error in https://github.com/openvinotoolkit/nncf/issues/3230#issuecomment-2630680904 must be defined elsewhere.

I intend to recreate the error reported above. What else can I do to investigate further?

**Comment by shumaari** [comment]
@daniil-lyakhov 

Apologies for the clutter above. I have created a draft pull request [3287](https://github.com/openvinotoolkit/nncf/pull/3287) . Could you please provide feedback?

Thank you!

**Comment by daniil-lyakhov** [comment]
@shumaari, thank you for your contribution! We are looking forward to collaborate with you in future!
